contracts/v2/core/WithdrawQueue.sol

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+import {Pausable} from "@openzeppelin/contracts/security/Pausable.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+contract WithdrawQueue is AccessControl, Pausable, ReentrancyGuard {
+    using Address for address payable;
+
+    struct Queue {
+        address to;
+        uint256 amount;
+    }
+
+    uint256 public constant CHUNK_SIZE = 32 ether;
+
+    uint256 front;
+    uint256 end;
+    uint256 lockedFront;
+    uint256 lockedAmount;
+
+    mapping(uint256 => Queue) public queue;
+
+    error NoAvailableQueue();
+
+    constructor() {
+        grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function push(address to) external payable nonReentrant {
+        Queue memory newWithdraw = Queue({to: to, amount: msg.value});
+
+        queue[end] = newWithdraw;
+        lockedAmount += msg.value;
+        end++;
+
+        if (lockedAmount >= CHUNK_SIZE) {
+            lockedAmount = 0;
+            lockedFront = end;
+        }
+    }
+
+    function processWithdraw() external nonReentrant {
+        if (lockedFront == front) {
+            revert NoAvailableQueue();
+        }
+
+        for (uint256 i = front; i < lockedFront; ) {
```

Comment:
would this work in the following example? 
10 users deposit, total requested is worth 320E 
9 users deposit, total to 32E, finally 1 user deposits, and is the rest of the Eth requested. 
Deployer currently has 200E they can return. 
So all users till the last user can be cleared/processed. The queue is partially cleared. 
When the deployer has more eth, they can process the final user, and any users after. 

But with this processWithdraw function, wouldn't it loop till the end, fail on the last user, and revert all previous sends and related accounting?

Reply:
@devlancer412 for viz ^